lib/backends/sdc/containers.js

```diff
@@ -559,60 +525,302 @@ function addRulesToFWAPI(opts, rules, payload, callback) {
+    var listParams = {
+        name: opts.config.externalNetwork || 'external',
+        fabric: false,
+        provisionable_by: opts.account.uuid
+    };
 
-    // No fabrics configured - fall back to provisioning on the external
-    // network
+    log.debug({ listParams: listParams },
+        format('Networks: fabrics not configured, using network %s',
+        listParams.name));
 
-    var netName = opts.config.externalNetwork || 'external';
-    listNetworks(opts, {name: netName}, function (err, networks) {
-        var external_net;
-        log.debug({err: err, res: networks}, 'list external networks');
+    listNetworks(opts, listParams, function (err, networks) {
+        log.debug({ err: err, res: networks },
+            format('Networks: listNetworks result for %s', listParams.name));
 
         if (err) {
-            callback(errors.napiErrorWrap(err, 'problem listing networks'));
+            callback(errors.napiErrorWrap(err,
+                format('Networks: problem listing network %s',
+                    listParams.name)));
+            return;
+        }
+
+        if (networks.length < 1) {
+            callback(new errors.DockerError(
+                format('unable to find %s network uuid', listParams.name)));
+            return;
+        }
+
+        payload.networks = [ { uuid: networks[0], primary: true} ];
+        return callback();
+    });
+}
+
+/*
+ * When fabrics are configured, but no specific network is supplied,
+ * or if 'bridge' is supplied, we will use the user's default fabric
+ * network, stored in UFDS.
+ *
+ * Additionally, if the user is publishing ports or we have the
+ * `fabricRequireExternal` option (currently used only by docker build,
+ * see DOCKER-705), we will also attach to configured network pool
+ * (typically the external/public pool).
+ */
+function defaultFabricNetwork(opts, container, payload, callback) {
+    var dc = opts.app.config.datacenterName;
+    var log = opts.log;
+    var requireExternal = false;
+
+    if (publishingPorts(container) || opts.fabricRequireExternal) {
+        requireExternal = true;
+    }
+
+    log.debug({ requireExternal: requireExternal },
+        'Networks: using default fabric network');
+
+    opts.app.ufds.getDcLocalConfig(opts.account.uuid, dc, function (err, conf) {
+        log.debug({err: err, conf: conf, account: opts.account.uuid},
+            'Networks: get DC local config');
+
+        if (err || !conf || !conf.defaultnetwork) {
+            callback(errors.ufdsErrorWrap(err,
+                'Networks: could not get default network'));
             return;
         }
 
-        networks.forEach(function (n) {
-            if (!external_net
-                && n.name === netName) {
+        payload.networks = [ {uuid: conf.defaultnetwork} ];
+
+        if (requireExternal) {
+            payload.networks.push({uuid: opts.config.overlay.externalPool});
+        }
+
+        payload.networks[payload.networks.length - 1].primary = true;
```

Comment:
Instead of making the last network to be added to the networks as primary, would it better to set the primary attribute specifically on either the external network? Something like:

```
if (requireExternal) {
    externalNetwork = {
      uuid: opts.config.overlay.externalPool,
      primary: true
    };
   
    payload.networks.push(externalNetwork);
}
```

It seems it would be more robust if the order of creation of network objects changes, or if a new network object gets added to the list of network as part of that function's implementation.